internal/mage/engine.go

```diff
@@ -48,3 +52,165 @@ func (t Engine) Lint(ctx context.Context) error {
+				buildkitBase = buildkitBase.WithFS(
+					buildkitBase.FS().WithFile("/usr/bin/"+sdkHelper+"-"+os+"-"+arch, helperBin),
+				)
+			}
+		}
+		platformVariants = append(platformVariants, buildkitBase)
+	}
+
+	return platformVariants
+}
+
+func (t Engine) Release(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	arches := []string{"amd64", "arm64"}
+	oses := []string{"linux", "darwin"}
+
+	imageRef, err := c.Container().Publish(ctx, engineImageRef, dagger.ContainerPublishOpts{
+		PlatformVariants: engineContainer(c, arches, oses),
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Println("Image published:", imageRef)
+
+	return nil
+}
+
+func localEngine(ctx context.Context, c *dagger.Client) (string, error) {
+	tmpfile, err := os.CreateTemp("", "dagger-engine-export")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpfile.Name())
+
+	_, err = c.Container().Export(ctx, tmpfile.Name(), dagger.ContainerExportOpts{
+		PlatformVariants: engineContainer(c, []string{runtime.GOARCH}, []string{runtime.GOOS}),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	containerName := "test-dagger-engine"
+	volumeName := "test-dagger-engine"
+	imageName := "localhost/test-dagger-engine:latest"
+
+	loadCmd := exec.CommandContext(ctx, "docker", "load", "-i", tmpfile.Name())
+	output, err := loadCmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("docker load failed: %w: %s", err, output)
+	}
+	_, imageID, ok := strings.Cut(string(output), "sha256:")
+	if !ok {
+		return "", fmt.Errorf("unexpected output from docker load: %s", output)
+	}
+	imageID = strings.TrimSpace(imageID)
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"tag",
+		imageID,
+		imageName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker tag: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"rm",
+		"-fv",
+		containerName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker rm: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"run",
+		"-d",
+		"--rm",
+		"-v", volumeName+":/var/lib/buildkit",
+		"--name", containerName,
+		"--privileged",
+		imageName,
+		"--debug",
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker run: %w: %s", err, output)
+	}
+
+	return containerName, nil
+}
+
+func (t Engine) StartLocal(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	containerName, err := localEngine(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("export DAGGER_HOST=docker-container://" + containerName)
+	return nil
+}
+
+func (t Engine) Test(ctx context.Context) error {
+	c, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stderr))
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	containerName, err := localEngine(ctx, c)
```

Comment:
related to the other question. Should we either:

- Update all tests using the Go SDK to spin up a local engine. OR
- Leave as is and let the Go SDK use the embedded Engine (for now)